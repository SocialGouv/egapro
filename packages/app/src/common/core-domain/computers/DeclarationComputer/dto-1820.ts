export const dto1820 = {
  "declaration-existante": { status: "creation" },
  entreprise: {
    type: "entreprise",
    tranche: "251:999",
    entrepriseDéclarante: {
      adresse: "105 RTE DE PARTHENAY 79100 SAINT-JEAN-DE-THOUARS",
      codeNaf: "82.92Z",
      codePostal: "79100",
      raisonSociale: "SOCOPLAN",
      siren: "627020373",
      commune: "SAINT-JEAN-DE-THOUARS",
      département: "79",
      région: "75",
    },
  },
  commencer: { annéeIndicateurs: 2022, siren: "627020373" },
  declarant: {
    accordRgpd: true,
    email: "demo@travail.fr",
    nom: "coll",
    prénom: "laet",
    téléphone: "0677614636",
  },
  "periode-reference": {
    périodeSuffisante: "oui",
    finPériodeRéférence: "2022-12-31",
    effectifTotal: 273,
  },
  remunerations: {
    estCalculable: "oui",
    mode: "niveau_branche",
    cse: "oui",
    dateConsultationCSE: "2023-02-22",
  },
  "remunerations-coefficient-branche": {
    catégories: [
      {
        nom: "",
        tranches: { ":29": 0.5, "30:39": 4, "40:49": 1.6, "50:": 2 },
      },
      {
        nom: "",
        tranches: { ":29": -0.8, "30:39": 5, "40:49": 2.6, "50:": 4.3 },
      },
      {
        nom: "",
        tranches: { ":29": "", "30:39": 10.5, "40:49": "", "50:": 8.6 },
      },
      {
        nom: "",
        tranches: { ":29": "", "30:39": "", "40:49": -55.6, "50:": -8.8 },
      },
    ],
  },
  "remunerations-resultat": {
    note: 37,
    populationFavorable: "femmes",
    résultat: 2.4,
  },
  augmentations: {
    estCalculable: "oui",
    populationFavorable: "hommes",
    résultat: 1.4,
    note: 20,
    catégories: { ouv: 0, emp: 1.4, tam: "", ic: "" },
  },
  promotions: {
    estCalculable: "oui",
    populationFavorable: "femmes",
    résultat: 4.7,
    note: 10,
    catégories: { ouv: 1.3, emp: -5.9, tam: "", ic: "" },
  },
  "conges-maternite": {
    estCalculable: "oui",
    résultat: 100,
    note: 15,
  },
  "hautes-remunerations": { résultat: 5, note: 10 },
};
