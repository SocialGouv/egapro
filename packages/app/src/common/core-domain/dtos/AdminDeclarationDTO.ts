export type AdminDeclarationDTO = {
  createdAt: Date;
  declarantEmail: string;
  declarantFirstName: string;
  declarantLastName: string;
  name: string;
  siren: string;
  type: "index" | "repeq";
  year: number;
} & (
  | {
      index: number;
      type: "index";
      ues?: {
        companies?: Array<{
          name: string;
          siren: string;
        }>;
        name: string;
      };
    }
  | {
      type: "repeq";
    }
);

export const columnMap = {
  type: "Type",
  siren: "SIREN",
  year: "Année",
  name: "Raison sociale",
  createdAt: "Date déclaration",
  declarantEmail: "Email",
  declarantFirstName: "Prénom",
  declarantLastName: "Nom",
  index: "Index",
  ues: "UES",
} as const satisfies Record<keyof Exclude<AdminDeclarationDTO, { type: "repeq" }>, string>;

export type OrderableColumn = Exclude<keyof typeof columnMap, "ues">;

export const orderableColumns = [
  "type",
  "siren",
  "year",
  "name",
  "createdAt",
  "declarantEmail",
  "declarantFirstName",
  "declarantLastName",
  "index",
] as const satisfies readonly OrderableColumn[];
