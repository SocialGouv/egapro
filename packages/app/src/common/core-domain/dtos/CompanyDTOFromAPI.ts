import { type COUNTIES, type CountryIsoCode, type NAF, type REGIONS } from "@common/dict";

export type CompanyDTOFromAPI = {
  address?: string;
  city?: string;
  cityCode?: string;
  countryIsoCode?: CountryIsoCode;
  /** Code d√©partement */
  countyCode?: keyof COUNTIES;
  nafCode?: keyof NAF;
  /* Raison sociale */
  name?: string;
  postalCode?: string;
  regionCode?: keyof REGIONS;
  siren?: string;
};
