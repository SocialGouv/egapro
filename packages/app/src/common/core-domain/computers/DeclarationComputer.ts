import { IndicateurCinqComputer } from "@common/core-domain/computers/IndicateurCinqComputer";
import { IndicateurDeuxComputer } from "@common/core-domain/computers/IndicateurDeuxComputer";
import { IndicateurDeuxTroisComputer } from "@common/core-domain/computers/IndicateurDeuxTroisComputer";
import { IndicateurQuatreComputer } from "@common/core-domain/computers/IndicateurQuatreComputer";
import { IndicateurTroisComputer } from "@common/core-domain/computers/IndicateurTroisComputer";
import { IndicateurUnComputer } from "@common/core-domain/computers/IndicateurUnComputer";
import { type Declaration } from "@common/core-domain/domain/Declaration";
import {
  type DeclarationDTO,
  type FavorablePopulationEnum,
  type IndicatorKey,
} from "@common/core-domain/dtos/DeclarationDTO";
import assert from "assert";

import { CompanyWorkforceRange } from "../domain/valueObjects/declaration/CompanyWorkforceRange";
import { FavorablePopulation } from "../domain/valueObjects/declaration/indicators/FavorablePopulation";

const indicateurUn = new IndicateurUnComputer();
const indicateurDeux = new IndicateurDeuxComputer(indicateurUn);
const indicateurTrois = new IndicateurTroisComputer(indicateurUn);
const indicateurDeuxTrois = new IndicateurDeuxTroisComputer(indicateurUn);
const indicateurQuatre = new IndicateurQuatreComputer();
const indicateurCinq = new IndicateurCinqComputer();

export const indicatorNoteMax: Record<IndicatorKey, number> = {
  remunerations: indicateurUn.getMaxNote(),
  augmentations: indicateurDeux.getMaxNote(),
  promotions: indicateurTrois.getMaxNote(),
  "augmentations-et-promotions": indicateurDeuxTrois.getMaxNote(),
  "conges-maternite": indicateurQuatre.getMaxNote(),
  "hautes-remunerations": indicateurCinq.getMaxNote(),
};

type IndicatorBase =
  | {
      isComputable: false;
    }
  | { favorablePopulation?: FavorablePopulationEnum; isComputable: true; result: number };

type DeclarationComputerInputBase = {
  highRemunerations: Omit<Extract<IndicatorBase, { isComputable: true }>, "isComputable">;
  maternityLeaves:
    | Extract<IndicatorBase, { isComputable: false }>
    | Omit<Extract<IndicatorBase, { isComputable: true }>, "favorablePopulation">;
  remunerations: IndicatorBase;
};

type DeclarationComputerInputFrom50To250 = DeclarationComputerInputBase & {
  range: CompanyWorkforceRange.Enum.FROM_50_TO_250;
  salaryRaisesAndPromotions:
    | Extract<IndicatorBase, { isComputable: false }>
    | (Extract<IndicatorBase, { isComputable: true }> & { employeesCountResult: number });
};

type DeclarationComputerInputFrom250ToInfinity = DeclarationComputerInputBase & {
  promotions: IndicatorBase;
  range: CompanyWorkforceRange.Enum.FROM_251_TO_999 | CompanyWorkforceRange.Enum.FROM_1000_TO_MORE;
  salaryRaises: IndicatorBase;
};

type DeclarationComputerInput = DeclarationComputerInputFrom50To250 | DeclarationComputerInputFrom250ToInfinity;

type DeclarationComputerOutput = {
  [K in
    | "maternityLeaves"
    | "promotions"
    | "remunerations"
    | "salaryRaises"
    | "salaryRaisesAndPromotions" as `${K}Score`]?: number;
} & {
  computablePoints: number;
  highRemunerationsScore: number;
  index?: number; // undefined means NC.
  points: number;
};

/** Helper to convert a 3 values favorable population in a 2 values with undefined */
export const toBinaryFavorablePopulation = (populationFavorable: FavorablePopulation.Enum) =>
  (populationFavorable === FavorablePopulation.Enum.EQUALITY
    ? undefined
    : populationFavorable) as FavorablePopulationEnum;

/** Helper to convert a 2 values favorable population with undefined in a 3 values */
export const toTernaryFavorablePopulation = (populationFavorable?: FavorablePopulationEnum) =>
  (!populationFavorable ? FavorablePopulation.Enum.EQUALITY : populationFavorable) as FavorablePopulation.Enum;

export const DeclarationComputerInputBuilder = {
  fromDeclaration(domain: Declaration) {
    // When périodeSuffisante is not "oui", we can't compute the index.
    assert(domain.sufficientPeriod, "sufficientPeriod must be true to compute the index");
    assert(domain.company.range, "entreprise.range must be set");
    assert(
      domain.remunerations && domain.maternityLeaves && domain.highRemunerations,
      "indicators must be set when sufficientPeriod is true",
    );

    // Handle indicator 2, 3 and 2&3.
    if (domain.company.range.getValue() === CompanyWorkforceRange.Enum.FROM_50_TO_250) {
      assert(domain.salaryRaisesAndPromotions, "indicators must be set when sufficientPeriod is true");
      if (domain.salaryRaisesAndPromotions?.notComputableReason === undefined) {
        assert(
          domain.salaryRaisesAndPromotions.favorablePopulation,
          "salaryRaisesAndPromotions.favorablePopulation must be set when computable",
        );
        assert(
          domain.salaryRaisesAndPromotions.result !== undefined,
          "salaryRaisesAndPromotions.result must be set when computable",
        );
        assert(
          domain.salaryRaisesAndPromotions.employeesCountResult !== undefined,
          "salaryRaisesAndPromotions.employeesCountResult must be set when computable",
        );
      }
    } else {
      assert(domain.salaryRaises && domain.promotions, "indicators must be set when sufficientPeriod is true");
      if (domain.salaryRaises?.notComputableReason === undefined) {
        assert(domain.salaryRaises.favorablePopulation, "salaryRaises.favorablePopulation must be set when computable");
        assert(domain.salaryRaises.result !== undefined, "salaryRaises.result must be set when computable");
      }
      if (domain.promotions?.notComputableReason === undefined) {
        assert(domain.promotions.favorablePopulation, "promotions.favorablePopulation must be set when computable");
        assert(domain.promotions.result !== undefined, "promotions.result must be set when computable");
      }
    }

    if (domain.remunerations.notComputableReason === undefined) {
      assert(domain.remunerations.favorablePopulation, "remunerations.favorablePopulation must be set when computable");
      assert(domain.remunerations.result !== undefined, "remunerations.result must be set when computable");
    }

    if (domain.maternityLeaves.notComputableReason === undefined) {
      assert(domain.maternityLeaves.result !== undefined, "maternityLeaves.result must be set when computable");
    }

    const input: DeclarationComputerInput = {
      remunerations:
        domain.remunerations.notComputableReason === undefined
          ? {
              favorablePopulation: toBinaryFavorablePopulation(domain.remunerations.favorablePopulation!.getValue()),
              isComputable: true,
              result: domain.remunerations.result!.getValue() || 0,
            }
          : {
              isComputable: false,
            },

      maternityLeaves:
        domain.maternityLeaves.notComputableReason === undefined
          ? {
              isComputable: true,
              result: domain.maternityLeaves.result!.getValue(),
            }
          : {
              isComputable: false,
            },
      highRemunerations: {
        favorablePopulation: toBinaryFavorablePopulation(domain.highRemunerations.favorablePopulation.getValue()),
        result: domain.highRemunerations.result.getValue(),
      },
      ...(domain.company.range.getValue() === CompanyWorkforceRange.Enum.FROM_50_TO_250
        ? {
            range: CompanyWorkforceRange.Enum.FROM_50_TO_250,
            salaryRaisesAndPromotions:
              domain.salaryRaisesAndPromotions!.notComputableReason === undefined // Unfortunately, previous assert is ignored by TS.
                ? {
                    isComputable: true,
                    favorablePopulation: toBinaryFavorablePopulation(
                      domain.salaryRaisesAndPromotions!.favorablePopulation!.getValue(),
                    ),
                    result: domain.salaryRaisesAndPromotions!.result!.getValue(),
                    employeesCountResult: domain.salaryRaisesAndPromotions!.employeesCountResult!.getValue(),
                  }
                : {
                    isComputable: false,
                  },
          }
        : {
            range: domain.company.range.getValue() as
              | CompanyWorkforceRange.Enum.FROM_251_TO_999
              | CompanyWorkforceRange.Enum.FROM_1000_TO_MORE,
            salaryRaises:
              domain.salaryRaises!.notComputableReason === undefined
                ? {
                    isComputable: true,
                    favorablePopulation: toBinaryFavorablePopulation(
                      domain.salaryRaises!.favorablePopulation!.getValue(),
                    ),
                    result: domain.salaryRaises!.result!.getValue(),
                  }
                : {
                    isComputable: false,
                  },
            promotions:
              domain.promotions!.notComputableReason === undefined
                ? {
                    isComputable: true,
                    favorablePopulation: toBinaryFavorablePopulation(
                      domain.promotions!.favorablePopulation!.getValue(),
                    ),
                    result: domain.promotions!.result!.getValue(),
                  }
                : {
                    isComputable: false,
                  },
          }),
    };

    return input;
  },

  fromDeclarationDTO(dto: DeclarationDTO): DeclarationComputerInput {
    // When périodeSuffisante is not "oui", we can't compute the index.
    assert(dto["periode-reference"]?.périodeSuffisante === "oui", "periode-reference.périodeSuffisante must be set");

    assert(dto.entreprise?.tranche, "entreprise.tranche must be set");
    assert(dto.remunerations?.estCalculable !== undefined, "remunerations.estCalculable must be set");
    assert(dto["conges-maternite"]?.estCalculable !== undefined, "conges-maternite.estCalculable must be set");

    if (dto.remunerations.estCalculable === "oui") {
      assert(dto["remunerations-resultat"]?.résultat !== undefined, "résultat must be set if indicator is computable");
    }

    if (dto.entreprise.tranche === CompanyWorkforceRange.Enum.FROM_50_TO_250) {
      assert(
        dto["augmentations-et-promotions"]?.estCalculable !== undefined,
        "augmentations-et-promotions.estCalculable must be set",
      );

      if (dto["augmentations-et-promotions"].estCalculable === "oui") {
        assert(
          dto["augmentations-et-promotions"].résultat !== undefined,
          "résultat must be set if indicator is computable",
        );
        assert(
          dto["augmentations-et-promotions"].résultatEquivalentSalarié !== undefined,
          "résultatEquivalentSalarié must be set if indicator is computable",
        );
      }
    } else {
      assert(dto.augmentations?.estCalculable !== undefined, "augmentations.estCalculable must be set");
      assert(dto.promotions?.estCalculable !== undefined, "promotions.estCalculable must be set");

      if (dto.augmentations.estCalculable === "oui") {
        assert(dto.augmentations.résultat !== undefined, "résultat must be set if indicator is computable");
      }
      if (dto.promotions.estCalculable === "oui") {
        assert(dto.promotions.résultat !== undefined, "résultat must be set if indicator is computable");
      }
    }

    if (dto["conges-maternite"].estCalculable === "oui") {
      assert(dto["conges-maternite"].résultat !== undefined, "résultat must be set if indicator is computable");
    }

    assert(dto["hautes-remunerations"]?.résultat !== undefined, "result must be set");

    const input: DeclarationComputerInput = {
      remunerations:
        dto.remunerations.estCalculable === "oui"
          ? {
              favorablePopulation: dto["remunerations-resultat"]!.populationFavorable,
              isComputable: true,
              result: dto["remunerations-resultat"]!.résultat,
            }
          : {
              isComputable: false,
            },
      maternityLeaves:
        dto["conges-maternite"].estCalculable === "oui"
          ? {
              isComputable: true,
              result: dto["conges-maternite"].résultat,
            }
          : {
              isComputable: false,
            },
      highRemunerations: {
        favorablePopulation: dto["hautes-remunerations"].populationFavorable,
        result: dto["hautes-remunerations"].résultat,
      },

      ...(dto.entreprise.tranche === CompanyWorkforceRange.Enum.FROM_50_TO_250
        ? {
            range: CompanyWorkforceRange.Enum.FROM_50_TO_250,
            salaryRaisesAndPromotions:
              dto["augmentations-et-promotions"]!.estCalculable === "oui"
                ? {
                    isComputable: true,
                    favorablePopulation: dto["augmentations-et-promotions"]!.populationFavorable,
                    result: dto["augmentations-et-promotions"]!.résultat,
                    employeesCountResult: dto["augmentations-et-promotions"]!.résultatEquivalentSalarié,
                  }
                : {
                    isComputable: false,
                  },
          }
        : {
            range: dto.entreprise.tranche as
              | CompanyWorkforceRange.Enum.FROM_251_TO_999
              | CompanyWorkforceRange.Enum.FROM_1000_TO_MORE,
            salaryRaises:
              dto.augmentations!.estCalculable === "oui"
                ? {
                    isComputable: true,
                    favorablePopulation: dto.augmentations!.populationFavorable,
                    result: dto.augmentations!.résultat,
                  }
                : {
                    isComputable: false,
                  },
            promotions:
              dto.promotions!.estCalculable === "oui"
                ? {
                    isComputable: true,
                    favorablePopulation: dto.promotions!.populationFavorable,
                    result: dto.promotions!.résultat,
                  }
                : {
                    isComputable: false,
                  },
          }),
    };

    return input;
  },
};

export const computeDeclarationIndex = (input: DeclarationComputerInput): DeclarationComputerOutput => {
  let points = 0,
    computablePoints = 0,
    salaryRaisesAndPromotionsScore,
    salaryRaisesScore,
    promotionsScore,
    maternityLeavesScore,
    remunerationsScore;

  if (input.remunerations.isComputable) {
    remunerationsScore = indicateurUn.computeNote(input.remunerations.result) || 0;
    points += remunerationsScore;
    computablePoints += indicateurUn.getMaxNote();
  }

  if (input.range === CompanyWorkforceRange.Enum.FROM_50_TO_250) {
    if (input.salaryRaisesAndPromotions.isComputable) {
      if (
        input.remunerations.isComputable &&
        input.remunerations.favorablePopulation !== input.salaryRaisesAndPromotions.favorablePopulation
      ) {
        salaryRaisesAndPromotionsScore = indicateurDeuxTrois.getMaxNote();
      } else {
        const percentScore = indicateurDeuxTrois.computeNote(input.salaryRaisesAndPromotions.result) || 0;
        const absoluteScore =
          indicateurDeuxTrois.computeNote(input.salaryRaisesAndPromotions.employeesCountResult) || 0;

        salaryRaisesAndPromotionsScore = Math.max(percentScore, absoluteScore);
      }
      points += salaryRaisesAndPromotionsScore;
      computablePoints += indicateurDeuxTrois.getMaxNote();
    }
  } else {
    if (input.salaryRaises.isComputable) {
      if (
        input.remunerations.isComputable &&
        input.remunerations.favorablePopulation !== input.salaryRaises.favorablePopulation
      ) {
        salaryRaisesScore = indicateurDeux.getMaxNote();
      } else {
        salaryRaisesScore = indicateurDeux.computeNote(input.salaryRaises.result) || 0;
      }
      points += salaryRaisesScore;
      computablePoints += indicateurDeux.getMaxNote();
    }
    if (input.promotions.isComputable) {
      if (
        input.remunerations.isComputable &&
        input.remunerations.favorablePopulation !== input.promotions.favorablePopulation
      ) {
        promotionsScore = indicateurTrois.getMaxNote();
      } else {
        promotionsScore = indicateurTrois.computeNote(input.promotions.result) || 0;
      }
      points += promotionsScore;
      computablePoints += indicateurTrois.getMaxNote();
    }
  }

  if (input.maternityLeaves.isComputable) {
    maternityLeavesScore = indicateurQuatre.computeNote(input.maternityLeaves.result) || 0;
    points += maternityLeavesScore;
    computablePoints += indicateurQuatre.getMaxNote();
  }

  const highRemunerationsScore = indicateurCinq.computeNote(input.highRemunerations.result) || 0;
  points += highRemunerationsScore;
  computablePoints += indicateurCinq.getMaxNote();

  return {
    highRemunerationsScore,
    maternityLeavesScore,
    promotionsScore,
    remunerationsScore,
    salaryRaisesAndPromotionsScore,
    salaryRaisesScore,
    points,
    computablePoints,
    index: computablePoints >= 75 ? Math.round((points / computablePoints) * 100) : undefined,
  };
};
