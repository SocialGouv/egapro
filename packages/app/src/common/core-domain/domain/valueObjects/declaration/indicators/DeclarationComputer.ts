import { IndicateurCinqComputer } from "@common/core-domain/computers/IndicateurCinqComputer";
import { IndicateurDeuxComputer } from "@common/core-domain/computers/IndicateurDeuxComputer";
import { IndicateurDeuxTroisComputer } from "@common/core-domain/computers/IndicateurDeuxTroisComputer";
import { IndicateurQuatreComputer } from "@common/core-domain/computers/IndicateurQuatreComputer";
import { IndicateurTroisComputer } from "@common/core-domain/computers/IndicateurTroisComputer";
import { IndicateurUnComputer } from "@common/core-domain/computers/IndicateurUnComputer";
import { type Declaration } from "@common/core-domain/domain/Declaration";
import { type DeclarationDTO, type IndicatorKey } from "@common/core-domain/dtos/DeclarationDTO";
import assert from "assert";

import { CompanyWorkforceRange } from "../CompanyWorkforceRange";
import { type FavorablePopulation } from "./FavorablePopulation";

export const indicatorNoteMax: Record<IndicatorKey, number> = {
  remunerations: IndicateurUnComputer.prototype.getMaxNote(),
  augmentations: IndicateurDeuxComputer.prototype.getMaxNote(),
  promotions: IndicateurTroisComputer.prototype.getMaxNote(),
  "augmentations-et-promotions": IndicateurDeuxTroisComputer.prototype.getMaxNote(),
  "conges-maternite": IndicateurQuatreComputer.prototype.getMaxNote(),
  "hautes-remunerations": IndicateurCinqComputer.prototype.getMaxNote(),
};

type IndicatorBase =
  | {
      isComputable: false;
    }
  | { favorablePopulation: FavorablePopulation.Enum; isComputable: true; result: number };

type DeclarationComputerInputBase = {
  highRemunerations: Omit<Extract<IndicatorBase, { isComputable: true }>, "isComputable">;
  maternityLeaves:
    | Extract<IndicatorBase, { isComputable: false }>
    | Omit<Extract<IndicatorBase, { isComputable: true }>, "favorablePopulation">;
  remunerations: IndicatorBase;
};

type DeclarationComputerInputFrom50To250 = DeclarationComputerInputBase & {
  range: CompanyWorkforceRange.Enum.FROM_50_TO_250;
  salaryRaisesAndPromotions: IndicatorBase;
};

type DeclarationComputerInputFrom250ToInfinity = DeclarationComputerInputBase & {
  promotions: IndicatorBase;
  range: CompanyWorkforceRange.Enum.FROM_251_TO_999 | CompanyWorkforceRange.Enum.FROM_1000_TO_MORE;
  salaryRaises: IndicatorBase;
};

type DeclarationComputerInput = DeclarationComputerInputFrom50To250 | DeclarationComputerInputFrom250ToInfinity;

type DeclarationComputerOutput = {
  [K in keyof DeclarationComputerInput]?: {
    score: number;
  };
} & {
  computablePoints: number;
  index?: number; // undefined means NC.
  points: number;
};

export const DeclarationComputerInputBuilder = {
  fromDeclaration(domain: Declaration) {
    // When périodeSuffisante is not "oui", we can't compute the index.
    assert(domain.sufficientPeriod, "sufficientPeriod must be true to compute the index");
    assert(domain.company.range, "entreprise.range must be set");
    assert(
      domain.remunerations && domain.maternityLeaves && domain.highRemunerations,
      "indicators must be set when sufficientPeriod is true",
    );

    // Handle indicator 2, 3 and 2&3.
    if (domain.company.range.getValue() === CompanyWorkforceRange.Enum.FROM_50_TO_250) {
      assert(domain.salaryRaisesAndPromotions, "indicators must be set when sufficientPeriod is true");
      if (domain.salaryRaisesAndPromotions?.notComputableReason === undefined) {
        assert(
          domain.salaryRaisesAndPromotions.favorablePopulation,
          "salaryRaisesAndPromotions.favorablePopulation must be set when computable",
        );
        assert(domain.salaryRaisesAndPromotions.result, "salaryRaisesAndPromotions.result must be set when computable");
      }
    } else {
      assert(domain.salaryRaises && domain.promotions, "indicators must be set when sufficientPeriod is true");
      if (domain.salaryRaises?.notComputableReason === undefined) {
        assert(domain.salaryRaises.favorablePopulation, "salaryRaises.favorablePopulation must be set when computable");
        assert(domain.salaryRaises.result, "salaryRaises.result must be set when computable");
      }
      if (domain.promotions?.notComputableReason === undefined) {
        assert(domain.promotions.favorablePopulation, "promotions.favorablePopulation must be set when computable");
        assert(domain.promotions.result, "promotions.result must be set when computable");
      }
    }

    if (domain.remunerations.notComputableReason === undefined) {
      assert(domain.remunerations.favorablePopulation, "remunerations.favorablePopulation must be set when computable");
      assert(domain.remunerations.result, "remunerations.result must be set when computable");
    }

    if (domain.maternityLeaves.notComputableReason === undefined) {
      assert(domain.maternityLeaves.result, "maternityLeaves.result must be set when computable");
    }

    const input: DeclarationComputerInput = {
      remunerations:
        domain.remunerations.notComputableReason === undefined
          ? {
              favorablePopulation: domain.remunerations.favorablePopulation!.getValue(),
              isComputable: true,
              result: domain.remunerations.result!.getValue() || 0,
            }
          : {
              isComputable: false,
            },

      maternityLeaves:
        domain.maternityLeaves.notComputableReason === undefined
          ? {
              isComputable: true,
              result: domain.maternityLeaves.result!.getValue(),
            }
          : {
              isComputable: false,
            },
      highRemunerations: {
        favorablePopulation: domain.highRemunerations.favorablePopulation.getValue(),
        result: domain.highRemunerations.result.getValue(),
      },
      ...(domain.company.range.getValue() === CompanyWorkforceRange.Enum.FROM_50_TO_250
        ? {
            range: CompanyWorkforceRange.Enum.FROM_50_TO_250,
            salaryRaisesAndPromotions:
              domain.salaryRaisesAndPromotions!.notComputableReason === undefined // Unfortunatelay, previous assert is ignored by TS.
                ? {
                    isComputable: true,
                    favorablePopulation: domain.salaryRaisesAndPromotions!.favorablePopulation!.getValue(),
                    result: domain.salaryRaisesAndPromotions!.result!.getValue(),
                  }
                : {
                    isComputable: false,
                  },
          }
        : {
            range: domain.company.range.getValue() as
              | CompanyWorkforceRange.Enum.FROM_251_TO_999
              | CompanyWorkforceRange.Enum.FROM_1000_TO_MORE,
            salaryRaises:
              domain.salaryRaises!.notComputableReason === undefined
                ? {
                    isComputable: true,
                    favorablePopulation: domain.salaryRaises!.favorablePopulation!.getValue(),
                    result: domain.salaryRaises!.result!.getValue(),
                  }
                : {
                    isComputable: false,
                  },
            promotions:
              domain.promotions!.notComputableReason === undefined
                ? {
                    isComputable: true,
                    favorablePopulation: domain.promotions!.favorablePopulation!.getValue(),
                    result: domain.promotions!.result!.getValue(),
                  }
                : {
                    isComputable: false,
                  },
          }),
    };

    return input;
  },

  fromDeclarationDTO(dto: DeclarationDTO): DeclarationComputerInput {
    // When périodeSuffisante is not "oui", we can't compute the index.
    assert(dto["periode-reference"]?.périodeSuffisante === "oui", "periode-reference.périodeSuffisante must be set");

    assert(dto.entreprise?.tranche, "entreprise.tranche must be set");
    assert(dto.remunerations?.estCalculable !== undefined, "remunerations.estCalculable must be set");
    assert(
      dto["augmentations-et-promotions"]?.estCalculable !== undefined,
      "augmentations-et-promotions.estCalculable must be set",
    );
    assert(dto.augmentations?.estCalculable !== undefined, "augmentations.estCalculable must be set");
    assert(dto.promotions?.estCalculable !== undefined, "promotions.estCalculable must be set");
    assert(dto["conges-maternite"]?.estCalculable !== undefined, "conges-maternite.estCalculable must be set");

    if (dto.remunerations.estCalculable === "oui") {
      assert(dto["remunerations-resultat"]?.résultat, "résultat must be set if indicator is computable");
      assert(
        dto["remunerations-resultat"]?.populationFavorable,
        "populationFavorable must be set if indicator is computable]",
      );
    }

    if (dto.entreprise.tranche === CompanyWorkforceRange.Enum.FROM_50_TO_250) {
      if (dto["augmentations-et-promotions"].estCalculable === "oui") {
        assert(dto["augmentations-et-promotions"].résultat, "résultat must be set if indicator is computable");
      }
    } else {
      if (dto.augmentations.estCalculable === "oui") {
        assert(dto.augmentations.résultat, "résultat must be set if indicator is computable");
      }
      if (dto.promotions.estCalculable === "oui") {
        assert(dto.promotions.résultat, "résultat must be set if indicator is computable");
      }
    }

    if (dto["conges-maternite"].estCalculable === "oui") {
      assert(dto["conges-maternite"].résultat, "résultat must be set if indicator is computable");
    }

    assert(dto["hautes-remunerations"]?.résultat, "result must be set");

    const input: DeclarationComputerInput = {
      remunerations:
        dto.remunerations.estCalculable === "oui"
          ? {
              favorablePopulation: dto["remunerations-resultat"]!.populationFavorable,
              isComputable: true,
              result: dto["remunerations-resultat"]!.résultat,
            }
          : {
              isComputable: false,
            },
      maternityLeaves:
        dto["conges-maternite"].estCalculable === "oui"
          ? {
              isComputable: true,
              result: dto["conges-maternite"].résultat,
            }
          : {
              isComputable: false,
            },
      highRemunerations: {
        favorablePopulation: dto["hautes-remunerations"].populationFavorable,
        result: dto["hautes-remunerations"].résultat,
      },

      ...(dto.entreprise.tranche === CompanyWorkforceRange.Enum.FROM_50_TO_250
        ? {
            range: CompanyWorkforceRange.Enum.FROM_50_TO_250,
            salaryRaisesAndPromotions:
              dto["augmentations-et-promotions"].estCalculable === "oui"
                ? {
                    isComputable: true,
                    favorablePopulation: dto["augmentations-et-promotions"].populationFavorable,
                    result: dto["augmentations-et-promotions"].résultat,
                  }
                : {
                    isComputable: false,
                  },
          }
        : {
            range: dto.entreprise.tranche as
              | CompanyWorkforceRange.Enum.FROM_251_TO_999
              | CompanyWorkforceRange.Enum.FROM_1000_TO_MORE,
            salaryRaises:
              dto.augmentations.estCalculable === "oui"
                ? {
                    isComputable: true,
                    favorablePopulation: dto.augmentations.populationFavorable,
                    result: dto.augmentations.résultat,
                  }
                : {
                    isComputable: false,
                  },
            promotions:
              dto.promotions.estCalculable === "oui"
                ? {
                    isComputable: true,
                    favorablePopulation: dto.promotions.populationFavorable,
                    result: dto.promotions.résultat,
                  }
                : {
                    isComputable: false,
                  },
          }),
    };

    return input;
  },
};

export const computeDeclarationIndex = (input: DeclarationComputerInput): DeclarationComputerOutput => {
  let points = 0,
    computablePoints = 0;

  if (input.remunerations.isComputable) {
    points += IndicateurUnComputer.prototype.computeNote(input.remunerations.result) || 0;
    computablePoints += IndicateurUnComputer.prototype.getMaxNote();
  }

  if (input.range === CompanyWorkforceRange.Enum.FROM_50_TO_250) {
    if (input.salaryRaisesAndPromotions.isComputable) {
      points += IndicateurDeuxTroisComputer.prototype.computeNote(input.salaryRaisesAndPromotions.result) || 0;
      computablePoints += IndicateurDeuxTroisComputer.prototype.getMaxNote();
    }
  } else {
    if (input.salaryRaises.isComputable) {
      points += IndicateurDeuxComputer.prototype.computeNote(input.salaryRaises.result) || 0;
      computablePoints += IndicateurDeuxComputer.prototype.getMaxNote();
    }
    if (input.promotions.isComputable) {
      points += IndicateurTroisComputer.prototype.computeNote(input.promotions.result) || 0;
      computablePoints += IndicateurTroisComputer.prototype.getMaxNote();
    }
  }

  if (input.maternityLeaves.isComputable) {
    points += IndicateurQuatreComputer.prototype.computeNote(input.maternityLeaves.result) || 0;
    computablePoints += IndicateurQuatreComputer.prototype.getMaxNote();
  }

  points += IndicateurCinqComputer.prototype.computeNote(input.highRemunerations.result) || 0;
  computablePoints += IndicateurCinqComputer.prototype.getMaxNote();

  return {
    points,
    computablePoints,
    index: computablePoints >= 75 ? Math.round((points / computablePoints) * 100) : undefined,
  };
};
