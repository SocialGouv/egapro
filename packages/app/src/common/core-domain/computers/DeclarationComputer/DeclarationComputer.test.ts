import { type DeclarationDTO } from "@common/core-domain/dtos/DeclarationDTO";

import { computeDeclarationIndex, DeclarationComputerInputBuilder } from "../DeclarationComputer";
import { dto1820 } from "./dto-1820";

it("DeclarationComputer: bug 1820 maternity leaves ", () => {
  const input = DeclarationComputerInputBuilder.fromDeclarationDTO(dto1820 as DeclarationDTO);

  const res = computeDeclarationIndex(input);

  expect(res).toMatchSnapshot();
});
