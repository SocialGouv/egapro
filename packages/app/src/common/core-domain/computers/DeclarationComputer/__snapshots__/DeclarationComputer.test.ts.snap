// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeclarationComputer: bug 1820 maternity leaves  1`] = `
{
  "computablePoints": 100,
  "highRemunerationsScore": 10,
  "index": 92,
  "maternityLeavesScore": 15,
  "points": 92,
  "promotionsScore": 10,
  "remunerationsScore": 37,
  "salaryRaisesAndPromotionsScore": undefined,
  "salaryRaisesScore": 20,
}
`;
